tasks/internal_reconcile_report_phantom/internal_reconcile_report_phantom.py

```diff
@@ -0,0 +1,204 @@
+import json
+from http import HTTPStatus
+from typing import Any, Dict, List, Union
+
+import fastjsonschema as fastjsonschema
+from cumulus_logger import CumulusLogger
+from fastjsonschema import JsonSchemaException
+from orca_shared.database import shared_db
+from orca_shared.database.shared_db import retry_operational_error
+from sqlalchemy import text
+from sqlalchemy.future import Engine
+
+OUTPUT_JOB_ID_KEY = "jobId"
+OUTPUT_ANOTHER_PAGE_KEY = "anotherPage"
+OUTPUT_PHANTOMS_KEY = "phantoms"
+PHANTOMS_COLLECTION_ID_KEY = "collectionId"
+PHANTOMS_GRANULE_ID_KEY = "granuleId"
+PHANTOMS_FILENAME_KEY = "filename"
+PHANTOMS_KEY_PATH_KEY = "keyPath"
+PHANTOMS_ORCA_ETAG_KEY = "orcaEtag"
+PHANTOMS_ORCA_LAST_UPDATE_KEY = "orcaLastUpdate"
+PHANTOMS_ORCA_SIZE_KEY = "orcaSize"
+
+LOGGER = CumulusLogger()
+
+PAGE_SIZE = 100
+
+
+def task(
+    job_id: int,
+    page_index: int,
+    db_connect_info: Dict[str, str],
+) -> Dict[str, Any]:
+    """
+    Args:
+        job_id: The unique ID of job/report.
+        page_index: The 0-based index of the results page to return.
+        db_connect_info: See requests_db.py's get_configuration for further details.
+    """
+    engine = shared_db.get_user_connection(db_connect_info)
+    granules = query_db(
+        engine,
+        job_id,
+        page_index,
+    )
+
+    return {
+        OUTPUT_JOB_ID_KEY: job_id,
+        OUTPUT_ANOTHER_PAGE_KEY: len(granules) > PAGE_SIZE,
+        OUTPUT_PHANTOMS_KEY: granules[0:PAGE_SIZE],  # we get one extra for anotherPage calculation.
+    }
+
+
+@retry_operational_error()
+def query_db(
+    engine: Engine,
+    job_id: str,
+    page_index: int,
+) -> List[Dict[str, Any]]:
+    """
+    Gets phantoms for the given job/page, up to PAGE_SIZE + 1 results.
+
+    Args:
+        engine: The sqlalchemy engine to use for contacting the database.
+        job_id: The unique ID of job/report.
+        page_index: The 0-based index of the results page to return.
+
+    Returns:
+        A list containing dicts matching the format of "phantoms" in output.json.
+    """
+    LOGGER.info(f"Retrieving page '{page_index}' of reports for job '{job_id}'")
+    with engine.begin() as connection:
+        sql_results = connection.execute(
+            get_phantoms_sql(),
+            [
+                {
+                    "job_id": job_id,
+                    "page_index": page_index,
+                    "page_size": PAGE_SIZE,
+                }
+            ],
+        )
+
+        phantoms = []
```

Comment:
There is nothing necessarily wrong with this approach but we could have the database do all this for us via SQL which may be more performant. By utilizing `array_to_json`, `array_agg`, and `row_to_json` in our PostgreSQL query we could return an array of JSON records without having to step through the records ourselves and put them into a dictionary. For this initial prototype the current implementation is fine. See the following reference for more information on how you could use the functions in SQL.

- https://hashrocket.com/blog/posts/faster-json-generation-with-postgresql

Consider creating a performance card and either add on to the one for the cursor performance mentioned earlier, or create a new one for looking into adjusting the SQL query for performance.

Reply:
https://bugs.earthdata.nasa.gov/browse/ORCA-419